frontend/javascripts/oxalis/model/accessors/accessor_helpers.js

```diff
@@ -0,0 +1,42 @@
+// @flow
+import _ from "lodash";
+
+/*
+  Wraps a given function so that it returns the same instance on consecutive
+  calls if the new and old return value are equal (given a equality function).
+
+  Motivation:
+  Most of our selectors are memoized which ensures that passing the same input,
+  produces the same outputs. As a result, instances are shared and shallow
+  comparisons minimize potential re-renders.
+  However, some selectors depend on the entire store state since they derive
+  complex computations. Every small change to the store, will produce a new
+  output even if it's equal to the last output.
+  So, instead of carefully decomposing all selectors to ensure maximum
+  instance-reusages, the `reuseInstanceOnEquality` can be used.
+
+  As a rule of thumb, this wrapper should be used for selectors which need
+  the entire store state.
+
+  Note that this function isn't of any use if the return type of the passed
+  function is a primitive value.
+ */
+export function reuseInstanceOnEquality<R, F: (...args: Array<any>) => R>(
```

Comment:
That's pretty cool! :1st_place_medal:

Reply:
You've only used this in one spot for now. How did you notice that this specific spot was a bottleneck? Performance tab measurement? There's probably other good candidates where this should be used :)

Reply:
This is a topic, I pondered on quite a bit. Since the new `ViewportStatusIndicator` uses a complex selector, I wanted to be very careful about its performance. I tried to measure the react rendering performance with the chrome performance tools and with the react profiler, but I couldn't get good insights, unfortunately. However, via logging I figured out that the `render` method was called with high frequency. That's why I wrote this helper function which fixes this. However, I'm not super sold on this solution. As I mentioned, in the above docstring, "carefully decomposing all selectors to ensure maximum instance-reusages" might be a better approach here (but also takes more effort).

Additionally, it might make sense to re-think how we listen to the store. Right now every store change will trigger countless subscriptions and each of them check whether their selector changed (due to memoization it might not be a big deal, but I have a bad feeling about this, since the subscribers might also do other things which trigger an unnecessary re-render). Instead, one could iterate over all used selectors and only trigger the updates where necessary. For example, if 10 subscribers listen on the same selector, one would only need to compare that selector once instead of 10 times. Not super sure how much this brings, but in combination with the selector-decomposition it seems at least intuitive to me that this should perform better.